Sources/NIONFS3/NFSTypes+Containers.swift

```diff
@@ -0,0 +1,529 @@
+            case 1: // AUTH_ERROR
+                let rawValue = try self.readNFSInteger(as: UInt32.self)
+                if let value = RPCAuthStatus(rawValue: rawValue) {
+                    return RPCReply(xid: xid, status: .messageDenied(.authError(value)))
+                } else {
+                    throw RPCErrors.illegalAuthStatus(rawValue)
+                }
+            default:
+                throw RPCErrors.illegalReplyRejectionStatus(rejectionKind)
+            }
+        default:
+            throw RPCErrors.illegalReplyStatus(acceptedOrDenied)
+        }
+    }
+
+    @discardableResult public mutating func writeRPCCall(_ call: RPCCall) -> Int {
+        return self.writeMultipleIntegers(
+            RPCMessageType.call.rawValue,
+            call.rpcVersion,
+            call.program,
+            call.programVersion,
+            call.procedure,
+            endianness: .big)
+        + self.writeRPCCredentials(call.credentials)
+        + self.writeRPCVerifier(call.verifier)
+    }
+
+    @discardableResult public mutating func writeRPCReply(_ reply: RPCReply) -> Int {
+        var bytesWritten = self.writeInteger(RPCMessageType.reply.rawValue, endianness: .big)
+
+        switch reply.status {
+        case .messageAccepted(_):
+            bytesWritten += self.writeInteger(0 /* accepted */, endianness: .big, as: UInt32.self)
+        case .messageDenied(_):
+            // FIXME: MSG_DENIED (spec name) isn't actually handled correctly here.
+            bytesWritten += self.writeInteger(1 /* denied */, endianness: .big, as: UInt32.self)
+        }
+        bytesWritten += self.writeInteger(0 /* verifier */, endianness: .big, as: UInt64.self)
+        + self.writeInteger(0 /* executed successfully */, endianness: .big, as: UInt32.self)
+        return bytesWritten
+    }
+
+
+    public mutating func readRPCCall(xid: UInt32) throws -> RPCCall {
+        guard let version = self.readInteger(endianness: .big, as: UInt32.self),
+              let program = self.readInteger(endianness: .big, as: UInt32.self),
+              let programVersion = self.readInteger(endianness: .big, as: UInt32.self),
+              let procedure = self.readInteger(endianness: .big, as: UInt32.self) else {
+                  throw NFS3Error.illegalRPCTooShort
+              }
+        let credentials = try self.readRPCCredentials()
+        let verifier = try self.readRPCVerifier()
+
+        guard version == 2 else {
+            throw RPCErrors.unknownVersion(version)
+        }
+
+        return RPCCall(xid: xid,
+                       rpcVersion: version,
+                       program: program,
+                       programVersion: programVersion,
+                       procedure: procedure,
+                       credentials: credentials,
+                       verifier: verifier)
+    }
+
+    public mutating func readNFSReply(programAndProcedure: RPCNFSProcedureID, rpcReply: RPCReply) throws -> RPCNFS3Reply {
```

Comment:
Should this return `nil` if there are insufficient bytes?

Reply:
same as above: no (because it's inside SUN RPC)